chunky/src/java/se/llbit/chunky/renderer/scene/ChunkLoader.java

```diff
@@ -0,0 +1,620 @@
+ * Loads chunks into octrees and other associated data structures.
+ */
+public class ChunkLoader {
+
+  private final String octreeImplementation;
+  private final int yMin;
+  private final int yMax;
+  private final int yClipMin;
+  private final int yClipMax;
+  private final EmitterSamplingStrategy emitterSamplingStrategy;
+  private final int gridSize;
+
+  public ChunkLoader(String octreeImplementation, int yMin, int yMax, int yClipMin, int yClipMax,
+      EmitterSamplingStrategy emitterSamplingStrategy, int gridSize) {
+    this.octreeImplementation = octreeImplementation;
+    this.yMin = yMin;
+    this.yMax = yMax;
+    this.yClipMin = yClipMin;
+    this.yClipMax = yClipMax;
+    this.gridSize = gridSize;
+    this.emitterSamplingStrategy = emitterSamplingStrategy;
+  }
+
+  /**
+   * Result from loading chunks.
+   */
+  public static class ChunkLoadResult {
+
+    final BlockPalette blockPalette;
+    final Octree worldOctree;
+    final Octree waterOctree;
+    Grid emitterGrid;
+    final Collection<Entity> entities;
+    final Collection<Entity> actors;
+    final Map<PlayerEntity, JsonObject> profiles;
+    final Vector3i origin;
+    final WorldTexture foliageTexture;
+    final WorldTexture waterTexture;
+    final WorldTexture grassTexture;
+    final Collection<ChunkPosition> loadedChunks;
+
+    public ChunkLoadResult(String octreeImplementation, int requiredDepth,
+        EmitterSamplingStrategy emitterSamplingStrategy, int gridSize) {
+      blockPalette = new BlockPalette();
+      worldOctree = new Octree(octreeImplementation, requiredDepth);
+      waterOctree = new Octree(octreeImplementation, requiredDepth);
+      emitterGrid =
+          emitterSamplingStrategy.equals(EmitterSamplingStrategy.NONE) ? null : new Grid(gridSize);
+      entities = new LinkedList<>();
+      actors = new LinkedList<>();
+      profiles = new HashMap<>();
+      origin = new Vector3i();
+      foliageTexture = new WorldTexture();
+      waterTexture = new WorldTexture();
+      grassTexture = new WorldTexture();
+      loadedChunks = new HashSet<>();
+    }
+  }
+
+  /**
+   * Loads chunks into octrees and other associated data.
+   *
+   * @param world        world to load
+   * @param chunksToLoad chunks to load
+   * @param taskTracker  tracker for keeping track of progress
+   * @param loadActors   whether or not to load actors from the world
+   */
+  public ChunkLoadResult loadChunks(World world, Collection<ChunkPosition> chunksToLoad,
+      TaskTracker taskTracker, boolean loadActors) {
+    BlockBounds bounds = Scene.calculateBounds(chunksToLoad);
+    int requiredDepth = Scene.calculateOctreeRequiredDepth(bounds, yMax, yMin);
+
+    ChunkLoadResult result = new ChunkLoadResult(octreeImplementation, requiredDepth,
+        emitterSamplingStrategy, gridSize);
+
+    try (Task task = taskTracker.task("(1/6) Loading regions")) {
+      task.update(2, 1);
+      //TODO is this necessary? Are we intentionally mutating origin here?
```

Comment:
Yes it is… also you're this now mutates the *result* of this method, so that's fine.

Reply:
Wups, meant to follow up in the review not leave as a TODO.

Ok, good to know it was intentional. Since the original mutating method was used in two places, one where the origin change was used immediately and here where only the depth info was used, I was concerned it was a minor bug causes by all of the lovely lovely side effects.

I'll remove the TODO